plugin/metadata/provider.go

```diff
@@ -8,40 +38,62 @@ import (
 
 // Provider interface needs to be implemented by each plugin willing to provide
 // metadata information for other plugins.
-// Note: this method should work quickly, because it is called for every request
-// from the metadata plugin.
 type Provider interface {
-	// List of variables which are provided by current Provider. Must remain constant.
-	MetadataVarNames() []string
-	// Metadata is expected to return a value with metadata information by the key
-	// from 4th argument. Value can be later retrieved from context by any other plugin.
-	// If value is not available by some reason returned boolean value should be false.
-	Metadata(ctx context.Context, state request.Request, variable string) (interface{}, bool)
+	// Metadata adds metadata to the context and returns a (potentially) new context.
+	// Note: this method should work quickly, because it is called for every request
+	// from the metadata plugin.
+	Metadata(ctx context.Context, state request.Request) context.Context
 }
 
-// M is metadata information storage.
-type M map[string]interface{}
+// Func is the type of function in the metadata, when called they return the value of the label.
+type Func func() string
 
-// FromContext retrieves the metadata from the context.
-func FromContext(ctx context.Context) (M, bool) {
-	if metadata := ctx.Value(metadataKey{}); metadata != nil {
-		if m, ok := metadata.(M); ok {
-			return m, true
+// Labels returns all metadata keys stored in the context. These label names should be named
+// as: plugin/NAME, where NAME is something descriptive.
+func Labels(ctx context.Context) []string {
+	if metadata := ctx.Value(key{}); metadata != nil {
+		if m, ok := metadata.(md); ok {
+			return keys(m)
 		}
 	}
-	return M{}, false
+	return nil
 }
 
-// Value returns metadata value by key.
-func (m M) Value(key string) (value interface{}, ok bool) {
-	value, ok = m[key]
-	return value, ok
+// ValueFunc returns the value function of label. If none can be found nil is returned. Calling the
+// function returns the value of the label.
+func ValueFunc(ctx context.Context, label string) Func {
```

Comment:
if we set the cache in the context (automatically). Then this method is back to "Value", and the caller do not care to call or not the fetching function.